### PR TITLE
Personal/namalu/implement connection protocol

### PR DESCRIPTION
**Pull request by namalu**
Implemented Shell Auth Service and Soda Connection.

### CODE REVIEW

File: MHVLib/Authentication/MHVShellAuthService.m [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    dispatch_async(self.authQueue, ^
+    {
+        if (!completion)
+        {
+            return;
+        }
+        
+        if (self.isAuthInProgress)
+        {
+            completion(nil, [NSError error:[NSError MHVOperationCannotBePerformed] withDescription:@"Another authentication operation is currenlty running."]);
```
Comment (miburf) [inline MHVLib/Authentication/MHVShellAuthService.m]: nit, currently

File: MHVLib/Authentication/MHVShellAuthService.m [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+#pragma mark - Private
+
+- (NSString *)mraBooleanString
+{
+    return self.configuration.isMultiRecordApp ? @"true" : @"false";
+}
+
+- (NSString *)miaParamString
```
Comment (miburf) [inline MHVLib/Authentication/MHVShellAuthService.m:188]: aibParamString?  this and mraBooleanString seem weird.  The param strings could be built as a dictionary, then a new [dictionary queryString] method that converts.

File: MHVLib/Authentication/MHVShellAuthService.m [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+{
+    return self.configuration.isMultiRecordApp ? @"&aib=true" : @"";
+}
+
+- (NSString *)InstanceIdFromUrl:(NSURL *)url
```
Comment (miburf) [inline MHVLib/Authentication/MHVShellAuthService.m]: nit, instanceId

File: MHVLib/Configuration/MHVConfiguration.h [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+@property (nonatomic, assign) NSTimeInterval requestTimeoutDuration;
+
+/**
+ Gets the request time to live.
+ @note This property defines the "msg-ttl" in the HealthVault request header XML. It determines how long the same XML can be used before HealthVault determines the request invalid. This property corresponds to the "HV_DefaultRequestTimeToLiveSeconds" configuration value when reading from web.config. The value defaults to 1800 seconds.
+ */
+@property (nonatomic, assign) NSTimeInterval requestTimeToLiveDuration;
+
+/**
+ Gets the number of retries the .NET APIs will make when getting an internal error response (error 500) from HealthVault.
```
Comment (miburf) [inline MHVLib/Configuration/MHVConfiguration.h]: .NET :)

File: MHVLib/Configuration/MHVConfiguration.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+ */
+@property (nonatomic, strong) NSArray<NSUUID *> *supportedTypeVersions;
```
Comment (miburf) [inline MHVLib/Configuration/MHVConfiguration.h]: This doesn't sounds like it applies to app SDKs

File: MHVLib/Configuration/MHVConfiguration.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
```
Comment (miburf) [inline MHVLib/Configuration/MHVConfiguration.h]: This doesn't sounds like it applies to app SDKs

File: MHVLib/Configuration/MHVConfiguration.m [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+- (NSURL *)ensureTrailingSlashOnUrl:(NSURL *)url
+{
+    NSString *urlString = [url absoluteString];
+    
+    if ([urlString hasSuffix:@"/"])
+    {
+        return url;
+    }
+    
+    return [[NSURL alloc] initWithString:[NSString stringWithFormat:@"urlString%@", @"/"]];
```
Comment (miburf) [inline MHVLib/Configuration/MHVConfiguration.m]: C# string formatter.

File: MHVLib/Connection/MHVSessionCredentialClient.m [hunk 8]
```diff
@@ -55,12 +55,11 @@ - (void)getSessionCredentialWithCompletion:(void (^_Nonnull)(MHVSessionCredentia
         return;
     }
     
+    MHVMethod *method = [MHVMethod createAuthenticatedSessionToken];
+    method.parameters = [self infoSection];
+    
     [self.connection executeMethod:MHVMethod.createAuthenticatedSessionToken
```
Comment (miburf) [inline MHVLib/Connection/MHVSessionCredentialClient.m]: should now pass new "method" variable

File: MHVLib/Connection/MHVSodaConnection.m [hunk 9]
```diff
@@ -61,53 +72,226 @@ - (instancetype)initWithConfiguration:(MHVConfiguration *)configuration
...
+            return;
+        }
+        
+        [self setConnectionPropertiesFromKeychain];
+        
+        if (!self.sessionCredential || !self.applicationCreationInfo)
+        {
+            [self finishAuthWithError:[NSError error:[NSError MHVUnauthorizedError] withDescription:@"Authorization required to perform this operation."]
+                           completion:completion];
+        }
```
Comment (miburf) [inline MHVLib/Connection/MHVSodaConnection.m]: should return; too.

File: MHVLib/Controllers/MHVBrowserAuthBroker.m [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+@implementation MHVBrowserAuthBroker
+
+- (instancetype)init
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _authQueue = dispatch_queue_create("MHVAppProvisionController.authQueue", DISPATCH_QUEUE_SERIAL);
```
Comment (miburf) [inline MHVLib/Controllers/MHVBrowserAuthBroker.m]: nit, MHVBrowserAuthBroker

File: MHVLib/HVMobile/MHVServiceRequest.m [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+- (void)writeHeader:(NSMutableString *)header forBody:(NSString *)body
+{
+    [header appendXmlElementStart:@"header"];
+    {
+        [self writeMethodHeaders:header];
+        [self writeRecordHeaders:header];
+        [self writeAuthSessionHeader:header];
+        [self writeStandardHeaders:header];
+        [self writeHashHeader:header forBody:body];
+    }
```
Comment (miburf) [inline MHVLib/HVMobile/MHVServiceRequest.m:106]: nit, {} aren't needed.  A few more below too.

**Review by miburf — APPROVED** [review]
A few things to fix before merging.

File: MHVLib/Authentication/MHVShellAuthService.m (hunk 2 above)
Comment (namalu) [reply MHVLib/Authentication/MHVShellAuthService.m:188]: The initialism actually stands for Multi-Instance Aware. I'm going to leave this as is for now, until we have something working to unblock everyone else. We can come back and make improvements later.

**Pull request merged**